### PR TITLE
fix(api): refactor cleanup + Kottke port dilation guard (#82, #83)

### LINKED ISSUE #82: Kottke dilation leaks PEC occupancy gradient into port cells
# rfx Issue: Kottke dilation leaks PEC occupancy gradient into port cells

## Summary

`_forward_from_materials` clears `pec_occupancy` at the port cell (line 13033), but Kottke's dilation step (`occ_dilated = max(f, roll(f,±1,...))`) picks up non-zero occupancy from neighboring cells — particularly the z+1 neighbor when a lumped/wire port sits 1 cell below a PEC patch. This makes AD gradients through `port_s11_freqs` wave decomposition incorrect (60x–223% mismatch vs finite differences), blocking reliable gradient-based PEC optimization near ports.

## Affected APIs

- `topology_optimize()` with `TopologyDesignRegion(material_fg="pec")` — `density_to_material_fields` emits `pec_occupancy=rho_p` which goes through the same `_forward_from_materials` → Kottke path
- `sim.forward(pec_occupancy_override=...)` — manual PEC occupancy for parametric optimization
- Any use of `pec_occupancy` with a lumped or wire port whose cells are within the Kottke dilation radius (1 cell) of non-zero occupancy cells

## Reproduction

Minimal setup: probe-fed microstrip patch antenna.

```python
from rfx import Simulation, Box, GaussianPulse
import jax, jax.numpy as jnp

F0 = 10e9; DX = 0.25e-3; CPML = 8
sim = Simulation(freq_max=F0*2, domain=(...), boundary="cpml", cpml_layers=CPML, dx=DX)
sim.add(Box(...), material="pec")       # ground plane
sim.add(Box(...), material="ro4003c")   # substrate (h=0.787mm = 3.1 cells)
sim.add_port((feed_x, cy, z_mid_sub), "ez", impedance=50.0,
             waveform=GaussianPulse(f0=F0, bandwidth=0.4))

# Smooth PEC mask for patch at z=patch_z (1 cell above port)
def smooth_mask(W, L):
    ...  # sigmoid in xyz, sharpness=2000
    return mask

def objective(params):
    W, L = params
    result = sim.forward(pec_occupancy_override=smooth_mask(W, L),
                         n_steps=1500, port_s11_freqs=jnp.array([F0]),
                         checkpoint=True)
    s11 = result.s_params[0]
    return jnp.real(s11)**2 + jnp.imag(s11)**2

# AD vs FD comparison
params = jnp.array([W0, L0])
loss, ad_grad = jax.value_and_grad(objective)(params)
fd_grad = central_difference(objective, params, delta=0.1e-3)
# Result: L gradient AD-FD mismatch = 223% (should be <10%)
```

## Root Cause Analysis

In `_forward_from_materials` (api.py):

1. **Line 13033:** Port cell cleared: `pec_occupancy_local = pec_occupancy_local.at[idx].set(0.0)` ✓
2. **Line 13241:** Kottke computed from `pec_occupancy_local`:
   ```python
   aniso_inv_eps_run = kottke_inv_eps_from_occupancy(pec_occupancy_local, ...)
   ```
3. **Inside `kottke_inv_eps_from_occupancy`** (geometry/smoothing.py line 923):
   ```python
   occ_dilated = max(f, roll(f,1,axis=0), roll(f,-1,axis=0), ..., roll(f,±1,axis=2))
   ```
   At the port cell: `f[port]=0` (cleared), but `f[port_z+1]` (patch cell, 1 cell above) has high occupancy → `occ_dilated[port] ≈ f[port_z+1] ≈ 1.0`
4. `jnp.maximum` in AD propagates gradient to whichever neighbor is the argmax → gradient of `occ[patch_z, feed_x, cy]` w.r.t. design parameters leaks into `inv_eps[port]`
5. Since `inv_eps[port]` directly affects the E-field update at the port cell, which determines V/I for S11, the leaked gradient dominates the physically correct gradient.

### Geometry constraint

For a probe-fed microstrip patch:
- Port at z_mid_sub (substrate midplane)
- Patch PEC at z=patch_z (substrate top surface)
- Gap = H_SUB/2 = 0.394mm = **1.6 cells** at dx=0.25mm → Kottke dilation (1 cell) bridges the gap

This is the standard microstrip patch geometry — the issue affects any topology/parametric optimization of a probe-fed patch.

## Evidence

Tested 8 configurations over 3 days. Key results:

| Config | AD-FD L error | S11 resonance? | Notes |
|--------|---------------|----------------|-------|
| smooth z + Kottke ON + no carve-out, dx=0.5mm | **5849%** | yes (-9.4 dB) | Original failure |
| smooth z + Kottke ON + 7-cell carve-out, dx=0.25mm | ~5% (L) | yes (-11.9 dB) | Carve-out blocks dilation but shifts resonance +18% |
| smooth z + Kottke ON + no carve-out, dx=0.25mm | **223%** | yes (-12.8 dB) | Improved by finer mesh but still large |
| smooth z + Kottke OFF, dx=0.25mm | n/a | no (>1.0) | Kottke required for 1-cell PEC |
| hard z + Kottke ON, dx=0.25mm | n/a | no (-6.4 dB) | Kottke needs smooth z transition |
| hard z + Kottke OFF, dx=0.25mm | n/a | no | No resonance at all |

The 7-cell carve-out (zeroing port cell + 6 face-neighbors in the occupancy array) fixes the gradient but creates a 1.25mm hole in the patch conductor, shifting resonance by +18%.

## Proposed Fix

In `_forward_from_materials`, after the existing port clearing (line 13033), also clear all cells within the Kottke dilation radius of each port cell:

```python
# Existing: clear port cell
idx = grid.position_to_index(pe.position)
if pec_occupancy_local is not None:
    pec_occupancy_local = pec_occupancy_local.at[idx[0], idx[1], idx[2]].set(0.0)

# Proposed: also clear Kottke dilation neighbors
# This prevents occ_dilated[port] from picking up nearby PEC occupancy
for di, dj, dk in [(1,0,0),(-1,0,0),(0,1,0),(0,-1,0),(0,0,1),(0,0,-1)]:
    ni = min(max(idx[0]+di, 0), grid.nx-1)
    nj = min(max(idx[1]+dj, 0), grid.ny-1)
    nk = min(max(idx[2]+dk, 0), grid.nz-1)
    pec_occupancy_local = pec_occupancy_local.at[ni, nj, nk].set(0.0)
```

### Concern

This creates a physical hole in the PEC structure near the port, which shifts resonance. A more sophisticated fix would modify the Kottke computation itself to exclude port-adjacent cells from the dilation max-pool, rather than zeroing their occupancy. This would preserve the PEC geometry while preventing gradient leakage.

```python
# Alternative: mask out port neighborhood from Kottke dilation
port_mask = jnp.zeros(grid.shape, dtype=bool)
for port_idx in all_port_indices:
    for di, dj, dk in [6 neighbors + center]:
        port_mask = port_mask.at[ni, nj, nk].set(True)

# In kottke_inv_eps_from_occupancy:
occ_dilated = jnp.where(port_mask, 0.0, max(f, roll(f,...)))
```

## Impact

This issue blocks **all PEC topology optimization near probe-fed ports** — the primary use case for antenna inverse design. `topology_optimize()` with `material_fg="pec"` will have incorrect gradients whenever the design region is within 1 cell of a lumped or wire port, which is the standard configuration for microstrip antenna optimization.

## Environment

- rfx version: 1.6.3
- JAX 0.4.30, float32
- Tested on RTX 4090 (24GB) and A6000 (48GB)
- Kottke path: `RFX_PEC_OCC_KOTTKE=1`

### LINKED ISSUE #83: api/ refactor cleanup: type hint mismatch + stale references
## Summary

Post-refactor cleanup for the api.py → rfx/api/ package split (Part B, Stages 0-4).

### MEDIUM: `_assemble_materials` type hint and docstring list 6 return values, actual return is 7

- `_compile.py:116`: type annotation declares 6-tuple
- `_compile.py:301`: actual return has 7 elements (missing `boundary_pec_shapes`)
- `_compile.py:121-124`: docstring also omits `boundary_pec_shapes`
- All callers unpack 7 correctly — no runtime bug, but misleading for IDE/type checkers

### LOW: Stale line-number references to old monolithic api.py

- `_execute.py:703`: references "api.py:2012"
- `_execute.py:1459`: references "api.py ~4268"

### LOW: `_port_sparameter_entries` defined in `_execute.py` but only called from `_preflight.py`

- `_execute.py:168`: definition
- `_preflight.py:78,178`: sole callers

## Plan

Self-patch on a branch, PR for review.

### PR DESCRIPTION
**Pull request by bk-squared**
## Summary
- Fix _assemble_materials type hint/docstring (7-tuple, not 6) — #83
- Update stale line-number references from pre-refactor api.py — #83
- Clear Kottke dilation neighbors around port cells — #82

## Kottke port dilation guard

Kottke occ_dilated = max(f, roll(f,+/-1,...)) picks up non-zero occupancy from cells adjacent to each port. For probe-fed patches (port 1 cell below PEC sheet), this leaks the patch occupancy gradient into the port inv_eps tensor.

Before fix: 60x-223% AD-vs-FD gradient mismatch on |S11|^2 objectives
After fix: Port cell + 6 face-neighbors zeroed in pec_occupancy_local before kottke_inv_eps_from_occupancy call

Closes #82, closes #83

### CODE REVIEW

**Comment by bk-squared** [comment]
Review pass: surgical port-aware Kottke dilation guard matches the Option A recommendation in docs/research_notes/20260516_s11_structural_blocker (predecessor handoff that motivated this fix). _compile.py type-hint update is consistent with the existing 7-tuple return (boundary_pec_shapes already populated in main, line 232/285/301; all callers unpack 7 elements). Smoke-imports OK on fix/api-refactor-cleanup. MERGEABLE/CLEAN. Merging.
